unittests/unittest_ip_resolver.cpp

```diff
@@ -0,0 +1,369 @@
+/*
+Copyright (c) 2018 Contributors as noted in the AUTHORS file
+
+This file is part of 0MQ.
+
+0MQ is free software; you can redistribute it and/or modify it under
+the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 3 of the License, or
+(at your option) any later version.
+
+0MQ is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <unity.h>
+#include "../tests/testutil.hpp"
+#include "../src/wrappers.hpp"
+
+#include <ip_resolver.hpp>
+
+//  Dummy if_nametoindex implementation to make test hardware-independant
+unsigned int zmq_if_nametoindex (const char *ifname_)
+{
+    static const char * dummy_interfaces[] = {
+        "lo0",
+        "eth0",
+        "eth1",
+    };
+    unsigned lut_len =
+        sizeof (dummy_interfaces) / sizeof (dummy_interfaces[0]);
+
+    for (unsigned i = 0; i < lut_len; i++) {
+        if (strcmp (dummy_interfaces[i], ifname_) == 0) {
+            //  The dummy index will be the position in the array + 1 (0 is
+            //  invalid)
+            return i + 1;
+        }
+    }
+
+    //  Not found
+    return 0;
+}
+
+//  Generate an invalid but well-defined 'ip_addr_t'. Avoids testing
+//  uninitialized values if the code is buggy.
+static zmq::ip_addr_t test_bad_addr (void)
+{
+    zmq::ip_addr_t addr;
+
+    memset (&addr, 0xba, sizeof (addr));
+
+    return addr;
+}
+
+static void validade_ipv6_addr (const zmq::ip_addr_t &addr_,
+                                const struct in6_addr &expected_addr_,
+                                uint16_t expected_port_ = 0,
+                                uint16_t expected_zone_ = 0)
+{
+    const sockaddr_in6 *ip6_addr = &addr_.ipv6;
+
+    TEST_ASSERT_EQUAL (AF_INET6, addr_.generic.sa_family);
+
+    int neq = memcmp (&ip6_addr->sin6_addr,
+                      &expected_addr_,
+                      sizeof (expected_addr_));
+
+    TEST_ASSERT_EQUAL (0, neq);
+    TEST_ASSERT_EQUAL (htons (expected_port_), ip6_addr->sin6_port);
+    TEST_ASSERT_EQUAL (expected_zone_, ip6_addr->sin6_scope_id);
+}
+
+static void validade_ipv4_addr (const zmq::ip_addr_t &addr_,
+                                const struct in_addr &expected_addr_,
+                                uint16_t expected_port_ = 0)
+{
+    const sockaddr_in *ip4_addr = &addr_.ipv4;
+
+    TEST_ASSERT_EQUAL (AF_INET, addr_.generic.sa_family);
+    TEST_ASSERT_EQUAL (expected_addr_.s_addr, ip4_addr->sin_addr.s_addr);
+    TEST_ASSERT_EQUAL (htons (expected_port_), ip4_addr->sin_port);
+}
+
+void test_bind_any (int ipv6_)
```

Comment:
This contains a lot of test cases, and should be split up, instead of chaining test cases, and reconfiguring the resolver in between.